pkg/datapath/loader/loader.go

```diff
@@ -196,6 +197,83 @@ func patchHostNetdevDatapath(ep datapath.Endpoint, objPath, dstPath, ifName stri
 	return hostObj.Write(dstPath, opts, strings)
 }
 
+func isObsoleteDev(dev string) bool {
+	// exclude devices we never attach to/from_netdev to.
+	for _, prefix := range defaults.ExcludedDevicePrefixes {
+		if strings.HasPrefix(dev, prefix) {
+			return false
+		}
+	}
+
+	// exclude devices that will still be managed going forward.
+	for _, d := range option.Config.GetDevices() {
+		if dev == d {
+			return false
+		}
+	}
+
+	return true
+}
+
+// removeObsoleteNetdevPrograms removes cil_to_netdev and cil_from_netdev from devices
+// that cilium potentially doesn't manage anymore after a restart, e.g. if the set of
+// devices in option.Config.GetDevices() changes between restarts.
+func removeObsoleteNetdevPrograms() error {
```

Comment:
Let's document this:

This code assumes that the agent was upgraded from a prior version while maintaining the same list of managed physical devices. This ensures that all tc bpf filters get replaced using the naming convention of the 'current' agent build. For example, before 1.13, most filters were named e.g. `bpf_host.o:[to-host]`, to be changed to `cilium-<device>` in 1.13, then to `cil_to_host-<device>` in 1.14. As a result, this function only cleans up filters following the current naming scheme.